### PR TITLE
Adding support for remotely-sourced zoo datasets

### PR DESCRIPTION
**Pull request by brimoor**
Adds support for loading zoo datasets whose download/preparation instructions are stored in a GitHub repository or publicly accessible URL! 🎉 

Example datasets published in this format:
- https://github.com/voxel51/coco-2017
- https://github.com/voxel51/caltech101

## Usage

```py
import fiftyone as fo
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset(
    "https://github.com/voxel51/coco-2017",
    split="validation",
    max_samples=50,
    shuffle=True,
)

dataset = foz.load_zoo_dataset("https://github.com/voxel51/caltech101")
```

Once downloaded, the rest of the zoo API (ie the `foz` module and `fiftyone zoo` CLI) "knows" about remotely-sourced datasets, and henceforward they may be referred to by either the `name` in their `fiftyone.yml` or by the source URL from which they were downloaded:

```shell
fiftyone zoo datasets list

# These are equivalent
fiftyone zoo datasets info voxel51/coco-2017
fiftyone zoo datasets info https://github.com/voxel51/coco-2017

# These are equivalent
fiftyone zoo datasets info voxel51/caltech101
fiftyone zoo datasets info https://github.com/voxel51/caltech101

fiftyone zoo datasets delete voxel51/coco-2017
fiftyone zoo datasets delete voxel51/caltech101
```

## Creating remotely-sourced datasets

This is extensively documented in the PR, but the basic ingredients to defining a remotely-sourced zoo dataset are a `fiftyone.yml` file:

```yaml
name: voxel51/coco-2017
type: dataset
author: The COCO Consortium
version: 1.0.0
url: https://github.com/voxel51/coco-2017
source: http://cocodataset.org/#home
license: https://cocodataset.org/#termsofuse
description: The COCO-2017 dataset
fiftyone:
  version: "*"
supports_partial_downloads: true
tags:
 - image
 - detection
 - segmentation
splits:
 - train
 - validation
 - test
size_samples:
 - train: 118287
 - test: 40670
 - validation: 5000
```

and an `__init__.py` file that contains a `download_and_prepare()` method:

```py
def download_and_prepare(dataset_dir, split=None, **kwargs):
    """Downloads the dataset and prepares it for loading into FiftyOne.

    Args:
        dataset_dir: the directory in which to construct the dataset
        split (None): a specific split to download, if the dataset supports
            splits. The supported split values are defined by the dataset's
            YAML file
        **kwargs: optional keyword arguments that your dataset can define to
            configure what/how the download is performed

    Returns:
        a tuple of

        -   ``dataset_type``: a ``fiftyone.types.Dataset`` type that the
            dataset is stored in locally, or None if the dataset provides
            its own ``load_dataset()`` method
        -   ``num_samples``: the total number of downloaded samples for the
            dataset or split
        -   ``classes``: a list of classes in the dataset, or None if not
            applicable
    """

    # Download files and organize them in `dataset_dir`
    ...

    # Define how the data is stored
    dataset_type = fo.types.ImageClassificationDirectoryTree
    dataset_type = None  # custom ``load_dataset()`` method

    # Indicate how many samples have been downloaded
    # May be less than the total size if partial downloads have been used
    num_samples = 10000

    # Optionally report what classes exist in the dataset
    classes = None
    classes = ["cat", "dog", ...]

    return dataset_type, num_samples, classes
```

and possibly a `load_dataset()` method:

```py
def load_dataset(dataset, dataset_dir, split=None, **kwargs):
    """Loads the dataset into the given FiftyOne dataset.

    Args:
        dataset: a :class:`fiftyone.core.dataset.Dataset` to which to import
        dataset_dir: the directory to which the dataset was downloaded
        split (None): a split to load. The supported values are
            ``("train", "validation", "test")``
        **kwargs: optional keyword arguments that your dataset can define to
            configure what/how the load is performed
    """

    # Load data into samples
    samples = [...]

    # Add samples to the dataset
    dataset.add_samples(samples)
```

## Summary by CodeRabbit

- **New Features**
	- Added a `delete` command to the CLI for removing local dataset copies.
	- Introduced support for remote datasets in the Dataset Zoo, allowing users to specify dataset names or URLs.
	- Added a new section in documentation for managing remotely-sourced datasets.

- **Documentation**
	- Enhanced clarity and usability of CLI documentation with refined command descriptions and examples.
	- Updated user guide to clarify the distinction between built-in and remote datasets, along with dependencies.

### CODE REVIEW

File: fiftyone/core/cli.py [hunk 1]
```diff
@@ -1984,71 +1976,83 @@ class DatasetZooInfoCommand(Command):
...
-        if name not in downloaded_datasets:
-            print("Dataset '%s' is not downloaded" % name)
-        else:
-            dataset_dir, info = downloaded_datasets[name]
+        if dataset_dir is not None:
             print(dataset_dir)
-            print("\n***** Dataset info *****")
-            print(info)
+        else:
+            print("Dataset '%s' is not downloaded" % name_or_url)
```
Comment (minhtuev) [inline fiftyone/core/cli.py:2024]: If the url is very long, this statement might look a little bit odd. Maybe we can put a cap on it
```
print(f"Dataset '{name_or_url[:20]}' is not downloaded")
```

File: fiftyone/zoo/datasets/__init__.py [hunk 2]
```diff
@@ -1252,6 +1453,290 @@ def _is_dataset_ready(self, dataset_dir, info, overwrite=False):
...
+        # sys.modules[module.__name__] = module
+        spec.loader.exec_module(module)
+        return module
+
+
+def _normalize_ref(url_or_gh_repo):
+    if etaw.is_url(url_or_gh_repo):
+        return url_or_gh_repo
+
+    return "https://github.com/" + url_or_gh_repo
```
Comment (minhtuev) [inline fiftyone/zoo/datasets/__init__.py:1625]: not sure if I follow the logic here, do we assume that the default host for remote datasets is github?

Comment (minhtuev) [reply fiftyone/zoo/datasets/__init__.py:1625]: ah got it now!

**Review by minhtuev — COMMENTED** [review]
Added some comments but otherwise looks good! Do you need help testing this PR?

File: fiftyone/zoo/datasets/__init__.py [hunk 3]
```diff
@@ -1252,6 +1453,290 @@ def _is_dataset_ready(self, dataset_dir, info, overwrite=False):
...
+            `download_and_prepare()` and/or `load_dataset()` methods
+    """
+
+    def __init__(self, dataset_dir, url=None, **kwargs):
+        d = _load_dataset_metadata(dataset_dir)
+
+        if url is not None:
+            url = _normalize_ref(url)
+        else:
+            url = d.get("url")
```
[the remark this answers is not in the document]
Comment (brimoor) [reply fiftyone/zoo/datasets/__init__.py:1485]: The `url` from the `fiftyone.yml` file is not actually used anywhere in the current implementation. The `url` that the user originally passed to `download_zoo_dataset()` is recorded in the dataset's `info.json` and that is passed into this method.

File: fiftyone/core/cli.py (hunk 1 above)
Comment (brimoor) [reply fiftyone/core/cli.py:2024]: Hmm, we don't have a precedent of truncating long strings in logs so I'd prefer consistency in this case 😄

**Comment by brimoor** [comment]
> Added some comments but otherwise looks good! Do you need help testing this PR?

Nah it should be good to go. Big delta but the code paths are pretty simple; just verified that the zoo API methods work for builtin/remote datasets with/without splits 😄

**Review by minhtuev — APPROVED** [review]

**Pull request merged**